psi4/src/psi4/libsapt_solver/usapt0.cc

```diff
@@ -2305,57 +2252,41 @@ void USAPT0::mp2_terms()
 
     // ==> Master Loop <== //
 
-    fseek(Aa_arf,0L,SEEK_SET);
-    fseek(Ba_asf,0L,SEEK_SET);
-    fseek(Ca_asf,0L,SEEK_SET);
-    fseek(Da_arf,0L,SEEK_SET);
-    fseek(Ab_arf,0L,SEEK_SET);
-    fseek(Bb_asf,0L,SEEK_SET);
-    fseek(Cb_asf,0L,SEEK_SET);
-    fseek(Db_arf,0L,SEEK_SET);
-    for (int astart = 0; astart < max(naa, nba); astart += maxa_a) {
-        int na_ablock = (astart + maxa_a >= naa ? naa - astart : maxa_a);
-        int nb_ablock = (astart + maxb_a >= nba ? nba - astart : maxb_a);
+    for (int aastart = 0, bastart = 0; aastart < max(naa, nba); aastart += maxa_a, bastart += maxb_a) {
+        int na_ablock = (aastart + maxa_a >= naa ? naa - aastart : maxa_a);
+        int nb_ablock = (aastart + maxb_a >= nba ? nba - aastart : maxb_a);
 
         if ( na_ablock > 0) {
-            fread(Aa_arp[0],sizeof(double),na_ablock*narQ,Aa_arf);
-            fread(Ba_asp[0],sizeof(double),na_ablock*nasQ,Ba_asf);
-            fread(Ca_asp[0],sizeof(double),na_ablock*nasQ,Ca_asf);
-            fread(Da_arp[0],sizeof(double),na_ablock*narQ,Da_arf);
+            dfh->fill_tensor("Aa_ar", Aa_ar, std::make_pair(aastart, aastart + na_ablock));
+            dfh->fill_tensor("Ba_as", Ba_as, std::make_pair(aastart, aastart + na_ablock));
+            dfh->fill_tensor("Ca_as", Ca_as, std::make_pair(aastart, aastart + na_ablock));
+            dfh->fill_tensor("Da_ar", Da_ar, std::make_pair(aastart, aastart + na_ablock));
         }
 
         if ( nb_ablock > 0) {
-            fread(Ab_arp[0],sizeof(double),nb_ablock*nbrQ,Ab_arf);
-            fread(Bb_asp[0],sizeof(double),nb_ablock*nbsQ,Bb_asf);
-            fread(Cb_asp[0],sizeof(double),nb_ablock*nbsQ,Cb_asf);
-            fread(Db_arp[0],sizeof(double),nb_ablock*nbrQ,Db_arf);
+            dfh->fill_tensor("Ab_ar", Ab_ar, std::make_pair(bastart, bastart + nb_ablock));
+            dfh->fill_tensor("Bb_as", Bb_as, std::make_pair(bastart, bastart + nb_ablock));
+            dfh->fill_tensor("Cb_as", Cb_as, std::make_pair(bastart, bastart + nb_ablock));
+            dfh->fill_tensor("Db_ar", Db_ar, std::make_pair(bastart, bastart + nb_ablock));
         }
 
 
-        fseek(Aa_bsf,0L,SEEK_SET);
-        fseek(Ba_brf,0L,SEEK_SET);
-        fseek(Ca_brf,0L,SEEK_SET);
-        fseek(Da_bsf,0L,SEEK_SET);
-        fseek(Ab_bsf,0L,SEEK_SET);
-        fseek(Bb_brf,0L,SEEK_SET);
-        fseek(Cb_brf,0L,SEEK_SET);
-        fseek(Db_bsf,0L,SEEK_SET);
-        for (int bstart = 0; bstart < max(nab, nbb); bstart += maxa_b) {
-            int na_bblock = (bstart + maxa_b >= nab ? nab - bstart : maxa_b);
-            int nb_bblock = (bstart + maxb_b >= nbb ? nbb - bstart : maxb_b);
+        for (int abstart = 0, bbstart = 0 ; abstart < max(nab, nbb); abstart += maxa_b, bbstart += maxb_b) {
```

Comment:
As you know DF_Helper fills tensors using tuple indexing.  However this loop did not previously keep track of the indexing for the `astart`.  This was intrinsically done for DFERI as it used `fread`, which moved the file pointer for each loop, but DF_Helper needs to keep track of the index.  The exit condition and nature of the loops is the same, i just kept track of one more variable.